deploy-cloudrun/src/cloudRun.ts

```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as core from '@actions/core';
+import { GaxiosResponse } from 'gaxios';
+import { google, run_v1 } from 'googleapis';
+import {
+  GoogleAuth,
+  JWT,
+  Compute,
+  UserRefreshClient,
+} from 'google-auth-library';
+import { Service } from './service';
+
+/**
+ * Available options to create the client.
+ *
+ * @param credentials GCP JSON credentials (default uses ADC).
+ * @param endpoint GCP endpoint (useful for testing).
+ */
+type ClientOptions = {
+  credentials?: string;
+  projectId?: string;
+};
+
+/**
+ * Wraps interactions with the Google Cloud Run API
+ *
+ * @param region Region of the GCP resource.
+ * @param opts list of ClientOptions
+ * @returns CloudRun client
+ */
+export class CloudRun {
+  readonly methodOptions = {
+    userAgentDirectives: [
+      {
+        product: 'github-actions-deploy-cloudrun',
+        version: '0.1.0',
+      },
+    ],
+  };
+
+  private run = google.run('v1');
+  readonly auth: GoogleAuth;
+  readonly parent: string;
+  authClient: any;
+
+  constructor(region: string, opts?: ClientOptions) {
+    let projectId = opts?.projectId;
+    if (
+      !opts?.credentials &&
+      (!process.env.GCLOUD_PROJECT ||
+        !process.env.GOOGLE_APPLICATION_CREDENTIALS)
+    ) {
+      throw new Error(
+        'No method for authentication. Set credentials in this action or export credentials from the setup-gcloud action',
+      );
+    }
+    // Instatiate Auth Client
+    // This method looks for the GCLOUD_PROJECT and GOOGLE_APPLICATION_CREDENTIALS
+    // environment variables.
+    this.auth = new google.auth.GoogleAuth({
+      scopes: ['https://www.googleapis.com/auth/cloud-platform'],
+    });
+    // Set credentials, if any.
+    if (opts?.credentials) {
+      let creds = opts?.credentials;
+      if (!opts?.credentials.trim().startsWith('{')) {
+        creds = Buffer.from(creds, 'base64').toString('utf8');
+      }
+      const jsonContent = JSON.parse(creds);
+      this.auth.jsonContent = jsonContent;
+
+      if (!projectId) {
+        projectId = jsonContent.project_id;
+        core.info('Setting project Id from credentials');
+      }
+    } else if (!projectId) {
```

Comment:
It might be clearer if the assumptions are stated explicitly within the conditional statement as opposed to transitively like this. Consider something like:
`if (!projectId && json.project_id) {}
elseif (!projectId && process.env.GCLOUD_PROJECT) {}
else {throw new Error()}`